src/OperationProcessor.ts

```diff
@@ -75,580 +17,160 @@ export interface OperationProcessor {
-
-    if (!earlier(parentOpInfo.timestamp, opInfo.timestamp)) {
-      return false;
-    }
+      // Any non-create needs a previous operation hash  ...
+      if (!operation.previousOperationHash) {
+        return false;
+      }
 
-    // TODO Perform:
-    // - operation number validation
+      // ... that should match the hash of the latest valid operation (previousOperation)
+      if (operation.previousOperationHash !== getOperationHash(previousOperation)) {
+        return false;
+      }
 
-    // Get the DID Document formed up until the parent operation.
-    const didDocument = await this.lookup(parentOpHash);
+      // The current did document should contain the public key mentioned in the operation ...
+      const publicKey = OperationProcessorImpl.getPublicKey(currentDidDocument, operation.signingKeyId);
+      if (!publicKey) {
+        return false;
+      }
 
-    // Fetch the public key to be used for signature verification.
-    const operation = await this.operationStore.lookup(opHash);
-    const publicKey = OperationProcessorImpl.getPublicKey(didDocument!, operation.signingKeyId);
+      // ... and the signature should verify
+      if (!(await operation.verifySignature(publicKey))) {
+        return false;
+      }
 
-    // Signature verification.
-    let verified = false;
-    if (publicKey) {
-      verified = await operation.verifySignature(publicKey);
+      return true;
     }
-    return verified;
-
-    return true;
   }
 
-  /**
-   * Invalidates the given operation by:
-   * 1. Removing its parent's reference to it in the next version map.
-   * 2. Sets the operation status to be 'Invalid'.
-   * @param opHash The hash of the operation to be invalidated.
-   */
-  private invalidatePreviouslyValidOperation (opHash: OperationHash) {
-    const opInfo = this.opHashToInfo.get(opHash)!;
-
-    if (opInfo.parent) {
-      this.nextVersion.delete(opInfo.parent);
+  // Update a document with a specified operation.
+  private getUpdatedDocument (didDocument: DidDocument | undefined, operation: Operation): DidDocument {
+    if (operation.type === OperationType.Create) {
+      return this.getInitialDocument(operation);
+    } else {
+      return Operation.applyJsonPatchToDidDocument(didDocument!, operation.patch!);
     }
-
-    opInfo.status = OperationStatus.Invalid;
   }
 
-  private async processDescendantsWaitingOn (opHash: OperationHash) {
-    const waitingDescendants = this.waitingDescendants.get(opHash);
-    if (waitingDescendants === undefined) {
-      return;
-    }
-
-    for (const descendantHash of waitingDescendants) {
-      const descendantInfo = this.opHashToInfo.get(descendantHash)!;
-      // assert: descendantInfo.status === Unvalidated and descendantInfo.missingAncestor === opHash
-      await this.processInternal(descendantHash, descendantInfo);
-    }
+  // Get the initial DID document for a create operation
+  private getInitialDocument (createOperation: Operation): DidDocument {
```

Comment:
rename to `getOriginalDocument`, as _original document_ is now the official term used in protocol spec.